service/history/queues/slice.go

```diff
@@ -344,17 +347,15 @@ func (s *SliceImpl) shrinkPredicate() {
 		return
 	}
 
-	if len(s.executableTracker.pendingPerNamespace) > shrinkPredicateMaxPendingNamespaces {
+	// TODO: this should be generic enough to shrink any predicate type, probably doesn't belong here.
+	pendingPerKey := s.executableTracker.pendingPerKey
+	if len(pendingPerKey) > shrinkPredicateMaxPendingKeys {
 		// only shrink predicate if there're few namespaces left
 		return
 	}
 
-	pendingNamespaceIDs := make([]string, 0, len(s.executableTracker.pendingPerNamespace))
-	for namespaceID := range s.executableTracker.pendingPerNamespace {
-		pendingNamespaceIDs = append(pendingNamespaceIDs, namespaceID.String())
-	}
-	namespacePredicate := tasks.NewNamespacePredicate(pendingNamespaceIDs)
-	s.scope.Predicate = tasks.AndPredicates(s.scope.Predicate, namespacePredicate)
+	minimalPredicate := s.grouper.Predicate(maps.Keys(pendingPerKey))
+	s.scope.Predicate = tasks.AndPredicates(s.scope.Predicate, minimalPredicate)
```

Comment:
now that I think about it, the `tasks.AndPredicates` looks unnecessary to me.

Reply:
Yeah, good catch!